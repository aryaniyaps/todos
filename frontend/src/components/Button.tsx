import { FC, ButtonHTMLAttributes, DetailedHTMLProps } from "react";
import { Spinner } from "./Spinner";

const sizeMap = {
    big: "py-2 px-6 text-sm rounded-lg",
    small: "px-2 py-1 text-sm rounded-md",
    tiny: "px-1 text-sm rounded-sm",
};

const colorMap = {
    accent: "bg-accent hover:bg-accent-hover disabled:bg-accent-disabled",
    primary: "bg-primary-800 hover:bg-primary-700 disabled:bg-accent-hover",
    transparent: "bg-transparent",
};

interface ButtonProps
    extends DetailedHTMLProps<
        ButtonHTMLAttributes<HTMLButtonElement>,
        HTMLButtonElement
    > {
    /** How big is the button? */
    size?: keyof typeof sizeMap;
    /** The color of the button */
    color?: keyof typeof colorMap;
    /** Is the button loading? */
    loading?: boolean;
    /** Does the button have transitions? */
    transition?: boolean;
}

export const Button: FC<ButtonProps> = ({
    children,
    disabled,
    loading,
    transition,
    className = "",
    color = "primary",
    size = "small",
    ...props
}) => {
    return (
        <button
            disabled={disabled || loading}
            data-testid="button"
            className={`flex ${sizeMap[size]} text-button
            ${transition ? "transition duration-200 ease-in-out" : ""} 
            ${colorMap[color]} font-semibold items-center justify-center 
            ${disabled ? "cursor-not-allowed" : ""} ${className}`}
            {...props}
        >
            <span className={loading ? "opacity-0" : `flex items-center`}>
                {children}
            </span>
            {loading ? (
                <span className="absolute">
                    <Spinner size={size === "small" ? "2" : "4"} />
                </span>
            ) : null}
        </button>
    );
};
