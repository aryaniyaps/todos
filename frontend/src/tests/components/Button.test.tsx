import { render } from "@testing-library/react";
import Button from "../../components/Button";

describe("Button component tests", () => {
    it("should render a button", () => {
        const { getByText } = render(<Button>sample label</Button>);
        const text = getByText("sample label");
        expect(text).toBeVisible();
    });

    it("should render a disabled button when loading", () => {
        const { getByText } = render(<Button loading>sample label</Button>);
        const button = getByText("sample label").closest("button");
        expect(button).toBeDisabled();
    });

    it("should render a disabled button when disabled", () => {
        const { getByText } = render(<Button disabled>sample label</Button>);
        const button = getByText("sample label").closest("button");
        expect(button).toBeDisabled();
    });

    it("should match snapshot", () => {
        const { getByTestId } = render(<Button>sample label</Button>);
        const button = getByTestId("button");
        expect(button).toMatchSnapshot();
    });
});
