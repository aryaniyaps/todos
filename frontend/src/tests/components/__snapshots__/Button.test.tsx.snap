// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button tests should match snapshot 1`] = `
<button
  class="flex py-2 px-6 text-md rounded-lg  text-primary-100 bg-primary-600 hover:bg-primary-500 disabled:bg-primary-500 font-semibold items-center justify-center  "
  data-testid="button"
>
  <span
    class="flex items-center"
  >
    sample label
  </span>
</button>
`;
